pkg/controller/master/node/promote_controller.go

```diff
@@ -246,12 +247,19 @@ func (h *PromoteHandler) setPromoteResult(job *batchv1.Job, node *corev1.Node, s
 // If the cluster doesn't need to be promoted, return nil
 func selectPromoteNode(nodeList []*corev1.Node) *corev1.Node {
 	var (
-		managementNumber     int
-		managementRoleNumber int
-		canPromoteNumber     int
-		promoteNode          *corev1.Node
+		managementNumber               int
+		managementRoleNumber           int
+		canPromoteNumber               int
+		promoteNode                    *corev1.Node
+		managementNodeTopologyZoneList []string
 	)
 
+	for _, node := range nodeList {
```

Comment:
For performance, and also to eliminate the "complex code", this function is better to re-design:

1. following check should be done asap, because most of the time, it has already reached the state
```
	// waiting for the other node completed
	if managementRoleNumber != managementNumber {
		return nil
	}

	// there is no need to promote if the spec number has been reached
	specManagementNumber := getSpecManagementNumber(len(nodeList))
	promoteNodeNumber := specManagementNumber - managementNumber
	if promoteNodeNumber <= 0 {
		return nil
	}
```

2. consider 2 times of iteration:
(1) first round, check mgmt node role & promopt node,   by the way, get the managementNodeTopologyZoneList
then check number, if reached, or has active promotion, return;  otherwise, go next

(2) second round, find the desired node, with check of topology

this way, the complex code will also be solved.

Reply:
@w13915984028 I totally agree on this. I rewrote the function like you suggested and it's much simpeler now. The only thing I don't know the reason of is why we are promoting the oldest node instead of the first in line?

Reply:
@joeyloman Selecting the oldest node ensures the predictability of the promote behavior of the node. The user can always know the promote order of the node, that is, the order in which they add the node to the cluster.

Reply:
When a function is complained as `Complex Method`, it means it's very difficult to make sure the correctness. Such method is hard to understand and maintain.

Another reason to pay attention to the performance, is that, `selectPromoteNode` is always called `OnNodeChanged` ,  there could be frequent `OnNodeChanged`, but for most of the time, there is noting related to node promotion (e.g, 3 management nodes,  100+ work nodes). The performance lose in `selectPromoteNode` deserves an optimization.

```
func (h *PromoteHandler) OnNodeChanged(key string, node *corev1.Node) (*corev1.Node, error) {
	if node == nil || node.DeletionTimestamp != nil {
		return node, nil
	}

	nodeList, err := h.nodeCache.List(labels.Everything())
	if err != nil {
		return nil, err
	}

	promoteNode := selectPromoteNode(nodeList)
	if promoteNode == nil {
		return node, nil
	}
```

Reply:
Yeah. Improving `selectPromoteNode` performance and reducing complexity while ensuring correctness is certainly great